### PR TITLE
first version

### PR DESCRIPTION
**Pull request by bvg33**

### CODE REVIEW

File: Dao/src/main/java/com/epam/ems/config/MySqlDataBaseConfig.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+@Configuration
+public class MySqlDataBaseConfig {
+    private static final String DRIVER = "com.mysql.jdbc.Driver";
+    private static final String URL = "jdbc:mysql://localhost:3306/EPAM";
+    private static final String USERNAME = "root";
+    private static final String PASSWORD = "as122592as";
+    private static final int MIN_IDLE = 5;
+    private static final int MAX_IDLE = 10;
+    private static final int MAX_OPEN_PREPARE_STATEMENTS = 100;
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/config/MySqlDataBaseConfig.java:18]: Вынести все в проперти файл

File: Dao/src/main/java/com/epam/ems/config/HSQLDBConfig.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.jdbc.datasource.embedded.EmbeddedDatabase;
+import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseBuilder;
+import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType;
+
+import javax.sql.DataSource;
+
+public class HSQLDBConfig {
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/config/HSQLDBConfig.java:11]: Напиши, зачем тебе этот конфиг?

File: Dao/src/main/java/com/epam/ems/connection/ConnectionPool.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+//NO NEED
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/connection/ConnectionPool.java:1]: Зачем тут такой "класс"?

File: Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    private QueryCreator queryCreator;
+
+    private static final String NAME = "name";
+    private static final String DESCRIPTION = "description";
+    private static final String CREATE_DATE = "create_date";
+    private static final String LAST_UPDATE_DATE = "last_update_date";
+    private static final String PRICE = "price";
+    private static final String DURATION = "duration";
+    private static final String SORT_BY_NAME="sortByName";
+    private static final String SORT_BY_DATE="sortByDate";
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java:81]: Это дупликация есть в других классах, создай отдельный класс с константами или ENUM с одним строковым полем.

File: Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@Repository
+public class CertificateDaoImpl implements Dao<Certificate> {
+    private static JdbcTemplate TEMPLATE;
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java:32]: Why "static"?

File: Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    private static final String PRICE = "price";
+    private static final String DURATION = "duration";
+    private static final String SORT_BY_NAME="sortByName";
+    private static final String SORT_BY_DATE="sortByDate";
+    @Override
+    public Certificate getById(int id) throws DaoException {
+        try {
+            return TEMPLATE.query(GET_CERTIFICATES_BY_ID, new Object[]{id}, certificateRowMapper).get(0);
+        }catch (Exception e){
+            throw new DaoException("there is no certificate with id="+id);
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java:87]: Add whitespaces 
[
id=" + id
]

File: Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+                ps.setString(2, fields.get(DESCRIPTION));
+                ps.setString(3, fields.get(PRICE));
+                ps.setString(4, fields.get(DURATION));
+                ps.setString(5, fields.get(CREATE_DATE));
+                ps.setString(6, fields.get(LAST_UPDATE_DATE));
+                return ps;
+            }, keyHolder);
+        }
+        List<Integer> tagIds = getTagsIds(item);
+        if (tagIds != null) {
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dao/certificatedao/CertificateDaoImpl.java:117]: Change the following code with java internal functions: tagIds != null
[
nonNull(tagIds)
]
Check this everywhere in the code

File: Web/src/main/java/com/epam/ems/controllers/TagsController.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @DeleteMapping("/{id}")
+    public String deleteTag(@PathVariable int id) throws DaoException {
+        service.deleteById(id);
+        return "Deleted successfully";
+    }
+
+    @PostMapping("/new")
+    public String createTag(@RequestParam MultiValueMap<String, String> allRequestParams) {
```
Comment (AndreiRohau) [inline Web/src/main/java/com/epam/ems/controllers/TagsController.java:36]: Create PUT|PATCH endpoint for "update" entity process

File: Web/src/main/java/com/epam/ems/controllers/ExceptionsHandler.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.context.request.WebRequest;
+
+@ControllerAdvice
+public class ExceptionsHandler {
+    @ExceptionHandler(DaoException.class)
+    public final ResponseEntity<Object> handleAllExceptions(Exception ex, WebRequest request) {
+        String details = ex.getLocalizedMessage();
+        ErrorResponse errorResponse = new ErrorResponse("error code =" + HttpStatus.INTERNAL_SERVER_ERROR + "12", details);
+        return new ResponseEntity<>(errorResponse, HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (AndreiRohau) [inline Web/src/main/java/com/epam/ems/controllers/ExceptionsHandler.java:17]: Where possible try not to use HttpStatus.INTERNAL_SERVER_ERROR just INTERNAL_SERVER_ERROR as static import.
Less code you write, easier for others to "read" and understand it.

File: Web/src/main/java/com/epam/ems/controllers/CertificateController.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+    @GetMapping("/{id}")
+    public Certificate getCertificateById(@PathVariable int id) throws Exception {
+        return service.getById(id);
+    }
+
+    @DeleteMapping("/{id}")
+    public String deleteCertificate(@PathVariable int id) throws DaoException {
+        service.deleteById(id);
+        return "Certificate was successfully deleted";
```
Comment (AndreiRohau) [inline Web/src/main/java/com/epam/ems/controllers/CertificateController.java:32]: Do not return simple String message as a result (use ResponseEntity<String>).
Look at the snippet below:

@GetMapping("/status")
ResponseEntity<String> statusCheck() {
    return new ResponseEntity<>(
      "Success", 
      HttpStatus.OK);
}

File: Web/src/main/java/com/epam/ems/controllers/CertificateController.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @PostMapping("/new")
+    public String addCertificate(@RequestParam MultiValueMap<String, String> allRequestParams) {
+        service.update(allRequestParams);
+        return "Certificate was successfully added";
+    }
+
+    @PatchMapping("/update/{id}")
+    public String updateCertificate(@PathVariable int id, @RequestParam MultiValueMap<String, String> allRequestParams) {
```
Comment (AndreiRohau) [inline Web/src/main/java/com/epam/ems/controllers/CertificateController.java:42]: why do you need POST for create\update and PATCH?

File: Dao/src/main/java/com/epam/ems/dao/tagdao/TagDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+import javax.sql.DataSource;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public class TagDaoImpl implements Dao<Tag> {
+
+    private static JdbcTemplate TEMPLATE;
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dao/tagdao/TagDaoImpl.java:21]: Why static?

File: Dao/src/main/java/com/epam/ems/dao/mapper/CertificateRowMapper.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+public class CertificateRowMapper implements ResultSetExtractor<List<Certificate>> {
+    private static final String ID = "id";
+    private static final String NAME = "name";
+    private static final String DESCRIPTION = "description";
+    private static final String CREATE_DATE = "create_date";
+    private static final String LAST_UPDATE_DATE = "last_update_date";
+    private static final String PRICE = "price";
+    private static final String DURATION = "duration";
+    private static final String TAG_ID = "TAG_ID";
+    private static final String TAG_NAME = "TAG_NAME";
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dao/mapper/CertificateRowMapper.java:24]: Duplication

File: Dao/src/main/java/com/epam/ems/dao/mapper/TagRowMapper.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.stereotype.Component;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+@Component
+public class TagRowMapper implements RowMapper<Tag> {
+    private static final String NAME = "tag_name";
+    private static final String ID = "id";
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dao/mapper/TagRowMapper.java:13]: Duplication

File: Dao/src/main/java/com/epam/ems/dao/tagdao/TagDaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    public List<Tag> getByTagName(String name) {
+        return TEMPLATE.query(GET_BY_NAME_QUERY, new Object[]{name}, tagRowMapper);
+    }
+
+    @Override
+    public List<Tag> getEntitiesSortedByParameter(String sortType, String param) {
+        if (param == SORT_BY_MAME_PARAM) {
+            return TEMPLATE.query(GET_ALL_SORTED_BY_NAME + sortType, tagRowMapper);
+        }
+        return new ArrayList<>();
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dao/tagdao/TagDaoImpl.java:74]: Refactor method to "return" only once (to have only one exit point in method).

File: Dao/src/main/java/com/epam/ems/dto/Certificate.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    public Certificate(int id, String name, String description, int price, int duration, String createDate, String lastUpdateDate, List<Tag> tags) {
+        this.id = id;
+        this.name = name;
+        this.description = description;
+        this.price = price;
+        this.duration = duration;
+        this.createDate = createDate;
+        this.lastUpdateDate = lastUpdateDate;
+        this.tags = new ArrayList<>(tags);
+    }
```
Comment (AndreiRohau) [inline Dao/src/main/java/com/epam/ems/dto/Certificate.java:36]: These are almost similar constructors. Use this() conctructor to remove duplications.

File: Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+@Component
+public class CertificateCreator implements Creator<Certificate> {
+
+    private static final String ID = "id";
+    private static final String NAME = "name";
+    private static final String DESCRIPTION = "description";
+    private static final String TAG = "tag";
+    private static final String PRICE = "price";
+    private static final String DURATION = "duration";
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java:21]: Duplication

File: Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+        String name = getStringField(NAME, allRequestParams);
+        String description = getStringField(DESCRIPTION, allRequestParams);
+        int price = getIntField(PRICE, allRequestParams);
+        int duration = getIntField(DURATION, allRequestParams);
+        List<Tag> tags = getTags(allRequestParams);
+        String currentDate = dateHandler.getCurrentDate();
+        if (id != 0) {
+            return new Certificate(id, name, description, price, duration, null, currentDate, tags);
+        }
+        return new Certificate(name, description, price, duration, currentDate, currentDate, tags);
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java:38]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    private int getIntField(String fieldName, MultiValueMap<String, String> allRequestParams) {
+        if (allRequestParams.containsKey(fieldName)) {
+            String stringField = allRequestParams.get(fieldName).get(0);
+            return Integer.parseInt(stringField);
+        }
+        return 0;
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java:46]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    private String getStringField(String fieldName, MultiValueMap<String, String> allRequestParams) {
+        if (allRequestParams.containsKey(fieldName)) {
+            return allRequestParams.get(fieldName).get(0);
+        }
+        return null;
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java:53]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    private List<Tag> getTags(MultiValueMap<String, String> allRequestParams) {
+        if (allRequestParams.containsKey(TAG)) {
+            List<Tag> tags = new ArrayList<>();
+            List<String> tagsNames = allRequestParams.get(TAG);
+            tagsNames.stream().forEach(tagName -> tags.add(new Tag(tagName)));
+            return tags;
+        }
+        return new ArrayList<>();
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/logic/creator/CertificateCreator.java:63]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/logic/creator/TagCreator.java [hunk 22]
```diff
@@ -0,0 +1,16 @@
+package com.epam.ems.logic.creator;
+
+import com.epam.ems.dto.Tag;
+import org.springframework.stereotype.Component;
+import org.springframework.util.MultiValueMap;
+
+@Component
+public class TagCreator implements Creator<Tag> {
+    private static final String NAME = "name";
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/logic/creator/TagCreator.java:9]: Duplication

File: Service/src/main/java/com/epam/ems/service/AbstractService.java [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public void deleteById(int id) throws DaoException {
+        dao.removeById(id);
+    }
+
+    protected String getFilteredList(String paramName, MultiValueMap<String, String> allRequestParams) {
+        if (allRequestParams.containsKey(paramName)) {
+            return allRequestParams.getFirst(paramName);
+        }
+        return new String();
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/service/AbstractService.java:41]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/service/AbstractService.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    protected List<T> findByPartOfName(MultiValueMap<String, String> allRequestParams) {
+        String parameter = getFilteredList(GET_BY_NAME_PART, allRequestParams);
+        if (!parameter.isEmpty()) {
+            return dao.getByNamePart(parameter);
+        } else {
+            return new ArrayList<>();
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/service/AbstractService.java:49]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/service/AbstractService.java [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+        }
+    }
+
+    protected List<T> findByTagNames(MultiValueMap<String, String> allRequestParams) {
+        String parameter = getFilteredList(GET_BY_TAG_NAME, allRequestParams);
+        if (!parameter.isEmpty()) {
+            return dao.getByTagName(parameter);
+        } else {
+            return new ArrayList<>();
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/service/AbstractService.java:58]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/service/certificate/CertificateService.java [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class CertificateService extends AbstractService<Certificate> {
+
+    private Dao<Certificate> dao;
+
+    private static final String SORT_BY_NAME_PARAM = "sortByName";
+    private static final String SORT_BY_DATE_PARAM = "sortByDate";
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/service/certificate/CertificateService.java:21]: Check for duplication

File: Service/src/main/java/com/epam/ems/service/certificate/CertificateService.java [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+        List<Certificate> foundByPartOfName = findByPartOfName(allRequestParams);
+        List<Certificate> foundedCertificates = new ArrayList<>();
+        fillListIfEmpty(foundByPartOfName, foundByTagNames,sortedByParam);
+        foundByTagNames.stream().filter(foundByPartOfName::contains).
+                forEach(foundedCertificates::add);
+        if (sortedByParam.isEmpty()) {
+            return foundedCertificates;
+        }
+        sortedByParam.retainAll(foundedCertificates);
+        return sortedByParam;
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/service/certificate/CertificateService.java:57]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/service/certificate/CertificateService.java [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+    private List<Certificate> getListSortedByParameter(MultiValueMap<String, String> allRequestParams) {
+        String sortType;
+        if (allRequestParams.containsKey(SORT_BY_NAME_PARAM)) {
+            sortType = getFilteredList(SORT_BY_NAME_PARAM, allRequestParams);
+            return dao.getEntitiesSortedByParameter(sortType, SORT_BY_NAME_PARAM);
+        } else if (allRequestParams.containsKey(SORT_BY_DATE_PARAM)) {
+            sortType = getFilteredList(SORT_BY_DATE_PARAM, allRequestParams);
+            return dao.getEntitiesSortedByParameter(sortType, SORT_BY_DATE_PARAM);
+        }
+        return new ArrayList<>();
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/service/certificate/CertificateService.java:75]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/service/tag/TagServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+        List<Tag> foundByPartOfName = findByPartOfName(allRequestParams);
+        List<Tag> foundedTags = new ArrayList<>();
+        fillListIfEmpty(foundByPartOfName, foundByTagNames,sortedByParam);
+        foundByTagNames.stream().filter(foundByPartOfName::contains).
+                forEach(foundedTags::add);
+        if (sortedByParam.isEmpty()) {
+            return foundedTags;
+        }
+        sortedByParam.retainAll(foundedTags);
+        return sortedByParam;
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/service/tag/TagServiceImpl.java:48]: Leave only 1 return statement!

File: Service/src/main/java/com/epam/ems/service/tag/TagServiceImpl.java [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    private List<Tag> getTagsSortedByNames(MultiValueMap<String, String> allRequestParams) {
+        String sortType = getFilteredList(SORT_BY_NAME_PARAM, allRequestParams);
+        if (!sortType.isEmpty()) {
+            return dao.getEntitiesSortedByParameter(sortType, SORT_BY_NAME_PARAM);
+        } else {
+            return new ArrayList<>();
```
Comment (AndreiRohau) [inline Service/src/main/java/com/epam/ems/service/tag/TagServiceImpl.java:56]: Leave only 1 return statement!

**Review by AndreiRohau — CHANGES_REQUESTED** [review]
Well done, left some comments, start adding changes, and feel free to ask questions or express concerns to me.

File: Dao/src/main/java/com/epam/ems/dao/extractor/CertificateFieldExtractor.java [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+        map.put(PRICE, priceString);
+        map.put(ID, idString);
+        return map;
+    }
+
+    private String intFieldToString(int field) {
+        if (field != 0) {
+            return String.valueOf(field);
+        }
+        return null;
```
Comment (AliakseiYermachonak) [inline Dao/src/main/java/com/epam/ems/dao/extractor/CertificateFieldExtractor.java:44]: don't return null without processing it on the higher level

**Pull request closed without merging**